### PR TITLE
Add v1 of oscilloscope

### PR DESCRIPTION
**Pull request by isherman**

### CODE REVIEW

**Comment by isherman** [comment]
This is a first pass at the design proposed in https://github.com/farm-ng/tractor/issues/88.

Screen Shot 2020-09-22 at 12 36 34 AM

It's ready for some real world use to prioritize next steps.

I'm pretty happy with the visualizer registry, and the ease of adding new visualizers. Performance w/ log files in the 100s of megabytes feels snappy. Streaming at 30FPS is working, for high-frequency numeric data at least. Responsive design is on the right track, but will need refinement based on feedback. 3D visualization is still an open question.

Some known issues:
- No support for streaming images yet (i.e. remote resource archive).
- In "Overlay" and "Grid" views, newest data is on the last page rather than the first.
- Plots don't resize in response to panel subdivision. Workaround is to resize the window. Fix is the [Resize Observer API](https://developer.mozilla.org/en-US/docs/Web/API/Resize_Observer_API).

Performance will be an ongoing concern.

- [ ] Merge latest proto definitions before merging

File: app/frontend/src/components/scope/visualizers/ApriltagDetectionsVisualizer.tsx [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    const canvasElement = canvasRef?.current;
+    if (imageElement && canvasElement) {
+      canvasElement.width = imageElement.clientWidth;
+      canvasElement.height = imageElement.clientHeight;
+    }
+  };
+
+  useEffect(
+    () =>
+      autorun(() => {
```
Comment (isherman) [inline app/frontend/src/components/scope/visualizers/ApriltagDetectionsVisualizer.tsx:53]: check whether this is necessary

File: app/frontend/src/models/TarReader.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
+const filenameMaxLength = 1024;
```
Comment (isherman) [inline app/frontend/src/models/TarReader.ts:1]: revisit

File: .vscode/react.code-snippets [hunk 3]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (ethanrublee) [inline .vscode/react.code-snippets:1]: this is cool, need to learn how to use that feature of vscode...

**Comment by isherman** [comment]
Latest commit supports logs with unknown types in them. (I'll add the latest types too, just would like to try to ensure this doesn't have to update in lockstep with on-robot protos).

I did have to restructure the .tar file you attached a bit to conform with the path structure the scope currently expects. I should document somewhere.
[mycalibration01-ian.tar.zip](https://github.com/farm-ng/tractor/files/5278527/mycalibration01-ian.tar.zip)

**Review by ethanrublee — APPROVED** [review]
This works pretty well on my end.  I like it!  Merge at will.

**Comment by isherman** [comment]
Thanks, going to merge this. Tracking TODOs in https://github.com/farm-ng/tractor/issues/106.

**Pull request merged**